ethergo/submitter/chain_queue.go

```diff
@@ -85,30 +92,41 @@
 
 		cq.bumpTX(gCtx, tx)
 	}
-    cq.updateOldTxStatuses(gCtx)
+	cq.updateOldTxStatuses(gCtx)
 
 	err = cq.g.Wait()
 	if err != nil {
 		return fmt.Errorf("error in chainPendingQueue: %w", err)
 	}
 
-	sort.Slice(cq.reprocessQueue, func(i, j int) bool {
-		return cq.reprocessQueue[i].Nonce() < cq.reprocessQueue[j].Nonce()
-	})
-
-	calls := make([]w3types.Caller, len(cq.reprocessQueue))
-	txHashes := make([]common.Hash, len(cq.reprocessQueue))
-	for i, tx := range cq.reprocessQueue {
-		calls[i] = eth.SendTx(tx.Transaction).Returns(&txHashes[i])
-	}
-
-	cq.storeAndSubmit(ctx, calls, span)
+	cq.storeAndSubmit(ctx, cq.reprocessQueue, span)
 
 	return nil
 }
 
 // storeAndSubmit stores the txes in the database and submits them to the chain.
-func (c *chainQueue) storeAndSubmit(ctx context.Context, calls []w3types.Caller, span trace.Span) {
+func (c *chainQueue) storeAndSubmit(ctx context.Context, reprocessQueue []db.TX, parentSpan trace.Span) {
+	sort.Slice(reprocessQueue, func(i, j int) bool {
+		return reprocessQueue[i].Nonce() < reprocessQueue[j].Nonce()
+	})
+
+	calls := make([]w3types.Caller, len(reprocessQueue))
+	txHashes := make([]common.Hash, len(reprocessQueue))
+	spanners := make([]trace.Span, len(reprocessQueue))
+
+	for i, tx := range reprocessQueue {
+		_, span := c.metrics.Tracer().Start(ctx, "chainPendingQueue.submit", trace.WithAttributes(util.TxToAttributes(tx.Transaction)...))
+		spanners[i] = span
+		defer func() {
+			// in case this span doesn't get ended automatically below, end it here.
+			if span.IsRecording() {
+				metrics.EndSpan(span)
+			}
+		}()
```

Comment:
You're right, but overhead is small and this is still the most efficient way to do this imo

Reply:
Out of curiosity,  is eprformance hit much worse than:

```go

type myCall func()

var deferredCalls []myCall

defer func() {
   for _, call := range deferredCalls {
     call()
   }
}

for i, tx := range reprocessQueue {
// do some stuff
deferredCalls = append(deferredCalls, func() {
  // Add a span
}
```

I can see a potential case where it would be (if the context outside of the function needed to be captured for each func(). But I'd imagine the func initialization is smart enough to only capture span so as long as I'm adding a span in both.

I remember this was an issue (com/golang/go#14939) but I beleive it was fixed in [go 1.14](https://go.dev/doc/go1.14#runtime)

Reply:
Transactiopn count is limited to 50, but I can/will make a note of this